### PR TITLE
PR to address remaining CUDA issues

### PR DESCRIPTION
**Pull request by bornaehsani**
This PR addresses the following issues: 

- Issue #304 : Added hb_mc_coordinate_get_relative/absolute function and moved hb_mc_coordinate_to_index and hb_mc_dimension_to_length to bsg_manycore_coordinate.h

- Issue #290 : Added hb_mc_device_memset() to CUDA API and also added test_memset to cuda regression test to check it's functionality

- Issue #282 : Changed pointer arguments to constant as much as possible

- Issue #278 : Changed hb_mc_device_init() to no longer take in dimensions and initialize device with all tiles, also added hb_mc_device_init_custom_dimensions to replace the former function, changed all cuda regression tests accordingly.

- Issue #273 : Removed all deprecated functions, source files and header files.

- Removed all functions that should not be accessed by the user from bsg_manycore_cuda header file and turned them static

- Reordered all functions in bsg_manycore_cuda header file to be in the order that the programmer is supposed to call them 

- Added proper initialize and exit function for the following:
         hb_mc_device_tile_groups_init/exit: Tile group list in the device
         hb_mc_device_manycore_init_exit
         hb_mc_device_mesh_init/exit
         hb_mc_device_program_init_exit
         hb_mc_program_allocator_init_exit
         hb_mc_tile_group_init_exit
         hb_mc_tile_group_kernel_init/exit

- Renamed hb_mc_grid_init to hb_mc_application_init/exit in CUDA API and changed cuda regression tests accordingly

Thanks to Max and Dustin for their comments that initiated this PR. 

F1 Passes on both v1.0.2 and v3.0.1 
Cosim passes on v1.0.2
Rebased and ready to merge. **Should be merged with PR # 82 in bsg_manycore**.

### CODE REVIEW

File: cl_manycore/regression/library/test_manycore_init.h [hunk 1]
```diff
@@ -1,11 +1,9 @@
 #pragma once
 #ifdef COSIM
```
Comment (drichmond) [inline cl_manycore/regression/library/test_manycore_init.h]: How did these #ifdef cosims sneak back in? I thought I removed all of them

Comment (bornaehsani) [reply cl_manycore/regression/library/test_manycore_init.h]: Got them!

**Comment by bornaehsani** [comment]
@drichmond Took care of the ifdef cosim issue. Rebased on top of master, regression tests pass on F1 both v1.0.2 and v3.0.3
Ready to merge if you're OK with it.

File: cl_manycore/libraries/bsg_manycore_cuda.cpp [hunk 2]
```diff
@@ -1250,13 +1591,18 @@ int hb_mc_device_free (hb_mc_device_t *device, eva_t eva) {
...
+ * @parma[in]  src           EVA address of source to be copied from
+ * @parma[in]  dst           EVA address of destination to be copied into
  * @param[in]  name          EVA address of dst
- * @param[in]  count         Size of buffer to be copied
+ * @param[in]  count         Size of buffer (number of words) to be copied
  * @param[in]  hb_mc_memcpy_kind         Direction of copy (HB_MC_MEMCPY_TO_DEVICE / HB_MC_MEMCPY_TO_HOST)
  * @return HB_MC_SUCCESS if succesful. Otherwise an error code is returned.
  */
-int hb_mc_device_memcpy (hb_mc_device_t *device, void *dst, const void *src, uint32_t count, enum hb_mc_memcpy_kind kind) {
+int hb_mc_device_memcpy (	hb_mc_device_t *device,
```
Comment (drichmond) [inline cl_manycore/libraries/bsg_manycore_cuda.cpp]: @bornaehsani can you remove tabs from the source files you edited, and tabs like here -- where they're between the parenthesis and the first argument? This is inconsistent with the rest of our source code

File: cl_manycore/libraries/bsg_manycore_cuda.cpp [hunk 3]
```diff
@@ -1296,58 +1642,51 @@ int hb_mc_device_memcpy (hb_mc_device_t *device, void *dst, const void *src, uin
...
+ * @return HB_MC_SUCCESS if succesful. Otherwise an error code is returned.
  */
-static hb_mc_epa_t hb_mc_tile_group_get_finish_signal_addr(hb_mc_tile_group_t *tg) { 
-	hb_mc_epa_t finish_addr = 	HB_MC_CUDA_HOST_FINISH_SIGNAL_BASE_ADDR
-					+ ((hb_mc_coordinate_get_y(tg->id) * hb_mc_dimension_get_x(tg->grid_dim)
-					+ hb_mc_coordinate_get_x(tg->id)) << 2); /* TODO: Hardcoded */
-	return finish_addr;
-}
-
+int hb_mc_device_memset (	hb_mc_device_t *device,
```
Comment (drichmond) [inline cl_manycore/libraries/bsg_manycore_cuda.cpp]: Likewise here

**Pull request merged**